File: framework/configstore/tables/mcp.go
package tables

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/bytedance/sonic"
	"github.com/maximhq/bifrost/core/schemas"
	"github.com/maximhq/bifrost/framework/encrypt"
	"gorm.io/gorm"
)

// TableMCPClient represents an MCP client configuration in the database
type TableMCPClient struct {
	ID                      uint               `gorm:"primaryKey;autoIncrement" json:"id"` // ID is used as the internal primary key and is also accessed by public methods, so it must be present.
	ClientID                string             `gorm:"type:varchar(255);uniqueIndex;not null" json:"client_id"`
	Name                    string             `gorm:"type:varchar(255);uniqueIndex;not null" json:"name"`
	IsCodeModeClient        bool               `gorm:"default:false" json:"is_code_mode_client"`         // Whether the client is a code mode client
	ConnectionType          string             `gorm:"type:varchar(20);not null" json:"connection_type"` // schemas.MCPConnectionType
	ConnectionString        *schemas.SecretVar `gorm:"type:text" json:"connection_string,omitempty"`
	StdioConfigJSON         *string            `gorm:"type:text" json:"-"`                              // JSON serialized schemas.MCPStdioConfig
	TLSConfigJSON           *string            `gorm:"type:text" json:"-"`                              // JSON serialized schemas.MCPTLSConfig
	ToolsToExecuteJSON      string             `gorm:"type:text" json:"-"`                              // JSON serialized []string
	ToolsToAutoExecuteJSON  string             `gorm:"type:text" json:"-"`                              // JSON serialized []string
	HeadersJSON             string             `gorm:"type:text" json:"-"`                              // JSON serialized map[string]string
	AllowedExtraHeadersJSON string             `gorm:"type:text" json:"-"`                              // JSON serialized []string
	IsPingAvailable         *bool              `gorm:"default:true" json:"is_ping_available,omitempty"` // Whether the MCP server supports ping for health checks
	// NeedsSessionStickiness: nil/false = per-call connection (the default
	// for newly created clients); true = persistent shared connection
	// (today's only behavior — every pre-existing row is explicitly
	// backfilled to true by the migration, so this default only applies to
	// clients created after this column existed). Ignored/always-true for
	// non-http connection types.
	//
	// Deliberately no `gorm:"default:..."` tag: a DB-level default would
	// make ADD COLUMN's fast-default mechanism return that value for every
	// pre-existing row immediately, so the migration's `WHERE
	// needs_session_stickiness IS NULL` backfill would never match anything.
	// Leaving the column plain-nullable is what lets the migration tell
	// "pre-existing, needs backfill" (NULL) apart from "explicitly set".
	NeedsSessionStickiness *bool  `json:"needs_session_stickiness,omitempty"`
	ToolPricingJSON        string `gorm:"type:text" json:"-"`                      // JSON serialized map[string]float64
	ToolSyncInterval       int    `gorm:"default:0" json:"tool_sync_interval"`     // Per-client tool sync interval in seconds (0 = use global, negative = disabled)
	ToolExecutionTimeout   int    `gorm:"default:0" json:"tool_execution_timeout"` // Per-client tool execution timeout in seconds (0 = use global from tool_manager_config)

	// Per-user OAuth: discovered tools persisted so they survive restart
	DiscoveredToolsJSON string `gorm:"type:text" json:"-"` // JSON serialized map[string]schemas.ChatTool
	ToolNameMappingJSON string `gorm:"type:text" json:"-"` // JSON serialized map[string]string

	// OAuth authentication fields
	AuthType      string            `gorm:"type:varchar(20);default:'headers'" json:"auth_type"`                         // "none", "headers", "oauth", "per_user_oauth", "per_user_headers", "token_exchange"
	OauthConfigID *string           `gorm:"type:varchar(255);index;constraint:OnDelete:CASCADE" json:"oauth_config_id"`  // Foreign key to oauth_configs.ID with CASCADE delete
	OauthConfig   *TableOauthConfig `gorm:"foreignKey:OauthConfigID;references:ID;constraint:OnDelete:CASCADE" json:"-"` // Gorm relationship

	// Per-user-headers schema: admin-declared list of header *names* that each
	// caller must supply. Empty/null for all other auth types. Used by both
	// the resolver (intersect with persisted user values) and by
	// utils.StaticConfigHeaders (strip from plugin-visible static headers).
	PerUserHeaderKeysJSON string `gorm:"type:text" json:"-"` // JSON serialized []string

	// Token-exchange configuration (audience/scopes + the exchange
	// application's client credentials) for auth_type='token_exchange'. NULL
	// for all other auth types. Carries a client secret, so it is encrypted
	// at rest like the other credential-bearing columns; the exchange
	// endpoint and grant shape come from the deployment's identity-provider
	// integration at exchange time.
	TokenExchangeJSON *string `gorm:"type:text" json:"-"` // JSON serialized schemas.MCPTokenExchangeConfig

	AllowOnAllVirtualKeys bool `gorm:"default:false" json:"allow_on_all_virtual_keys"` // Whether to allow the MCP client to run on all virtual keys
	Disabled              bool `gorm:"default:false" json:"disabled"`                  // Whether the client is intentionally disabled

	// PendingOAuthConfigJSON stashes the inline `oauth_config` block from
	// config.json for shared-OAuth MCP clients (auth_type='oauth') that have
	// not yet been authorized by an admin. NULL on UI-created clients and on
	// rows whose OAuth has already been authorized — cleared by the OAuth
	// callback once oauth_configs.status='authorized'.
	//
	// Deserialised into PendingOAuthConfig by AfterFind; serialised back by
	// BeforeSave. Wire-side it surfaces as `oauth_config` (UI form parity).
	//
	// The column name is pinned explicitly: GORM's naming strategy would
	// otherwise derive pending_o_auth_config_json from the Go field name,
	// breaking the add-column migration and every raw map update that
	// addresses the column as pending_oauth_config_json.
	PendingOAuthConfigJSON *string `gorm:"column:pending_oauth_config_json;type:text" json:"-"`

	// Config hash is used to detect the changes synced from config.json file
	// Every time we sync the config.json file, we will update the config hash
	ConfigHash string `gorm:"type:varchar(255);null" json:"config_hash"`

	EncryptionStatus string `gorm:"type:varchar(20);default:'plain_text'" json:"-"`

	CreatedAt time.Time `gorm:"index;not null" json:"created_at"`
	UpdatedAt time.Time `gorm:"index;not null" json:"updated_at"`

	// Virtual fields for runtime use (not stored in DB)
	StdioConfig               *schemas.MCPStdioConfig         `gorm:"-" json:"stdio_config,omitempty"`
	TLSConfig                 *schemas.MCPTLSConfig           `gorm:"-" json:"tls_config,omitempty"`
	ToolsToExecute            schemas.WhiteList               `gorm:"-" json:"tools_to_execute"`
	ToolsToAutoExecute        schemas.WhiteList               `gorm:"-" json:"tools_to_auto_execute"`
	Headers                   map[string]schemas.SecretVar    `gorm:"-" json:"headers"`
	AllowedExtraHeaders       schemas.WhiteList               `gorm:"-" json:"allowed_extra_headers"`
	ToolPricing               map[string]float64              `gorm:"-" json:"tool_pricing"`
	DiscoveredTools           map[string]schemas.ChatTool     `gorm:"-" json:"-"`
	DiscoveredToolNameMapping map[string]string               `gorm:"-" json:"-"`
	PerUserHeaderKeys         []string                        `gorm:"-" json:"per_user_header_keys"`
	TokenExchange             *schemas.MCPTokenExchangeConfig `gorm:"-" json:"token_exchange,omitempty"` // Runtime mirror of TokenExchangeJSON
	PendingOAuthConfig        *schemas.OAuth2Config           `gorm:"-" json:"oauth_config,omitempty"`   // Runtime mirror of PendingOAuthConfigJSON
}

// TableName sets the table name for each model
func (TableMCPClient) TableName() string { return "config_mcp_clients" }

// BeforeSave is a GORM hook that serializes runtime fields (stdio config, tools, headers,
// pricing) into JSON columns and encrypts the connection string and headers before writing
// to the database. Environment-variable-backed connection strings are not encrypted.
func (c *TableMCPClient) BeforeSave(tx *gorm.DB) error {
	if c.StdioConfig != nil {
		data, err := json.Marshal(c.StdioConfig)
		if err != nil {
			return err
		}
		config := string(data)
		c.StdioConfigJSON = &config
	} else {
		c.StdioConfigJSON = nil
	}

	if c.TLSConfig != nil {
		data, err := c.TLSConfig.MarshalForStorage()
		if err != nil {
			return err
		}
		config := string(data)
		c.TLSConfigJSON = &config
	} else {
		c.TLSConfigJSON = nil
	}

	if c.ToolsToExecute != nil {
		if err := c.ToolsToExecute.Validate(); err != nil {
			return fmt.Errorf("invalid tools_to_execute: %w", err)
		}
		data, err := json.Marshal(c.ToolsToExecute)
		if err != nil {
			return err
		}
		c.ToolsToExecuteJSON = string(data)
	} else {
		c.ToolsToExecuteJSON = "[]"
	}

	if c.ToolsToAutoExecute != nil {
		if err := c.ToolsToAutoExecute.Validate(); err != nil {
			return fmt.Errorf("invalid tools_to_auto_execute: %w", err)
		}
		data, err := json.Marshal(c.ToolsToAutoExecute)
		if err != nil {
			return err
		}
		c.ToolsToAutoExecuteJSON = string(data)
	} else {
		c.ToolsToAutoExecuteJSON = "[]"
	}

	if c.Headers != nil {
		headersToSerialize := make(map[string]string, len(c.Headers))
		for key, value := range c.Headers {
			if value.IsFromSecret() {
				headersToSerialize[key] = value.GetRawRef()
			} else {
				headersToSerialize[key] = value.GetValue()
			}
		}
		data, err := json.Marshal(headersToSerialize)
		if err != nil {
			return err
		}
		c.HeadersJSON = string(data)
	} else {
		c.HeadersJSON = "{}"
	}

	if c.AllowedExtraHeaders != nil {
		if err := c.AllowedExtraHeaders.Validate(); err != nil {
			return fmt.Errorf("invalid allowed_extra_headers: %w", err)
		}
		data, err := json.Marshal(c.AllowedExtraHeaders)
		if err != nil {
			return err
		}
		c.AllowedExtraHeadersJSON = string(data)
	} else {
		c.AllowedExtraHeadersJSON = "[]"
	}

	if c.ToolPricing != nil {
		data, err := json.Marshal(c.ToolPricing)
		if err != nil {
			return err
		}
		c.ToolPricingJSON = string(data)
	} else {
		c.ToolPricingJSON = "{}"
	}

	if c.DiscoveredTools != nil {
		data, err := json.Marshal(c.DiscoveredTools)
		if err != nil {
			return err
		}
		c.DiscoveredToolsJSON = string(data)
	}

	if c.DiscoveredToolNameMapping != nil {
		data, err := json.Marshal(c.DiscoveredToolNameMapping)
		if err != nil {
			return err
		}
		c.ToolNameMappingJSON = string(data)
	}

	if c.PerUserHeaderKeys != nil {
		data, err := json.Marshal(c.PerUserHeaderKeys)
		if err != nil {
			return err
		}
		c.PerUserHeaderKeysJSON = string(data)
	} else {
		c.PerUserHeaderKeysJSON = ""
	}

	if c.TokenExchange != nil {
		data, err := json.Marshal(c.TokenExchange)
		if err != nil {
			return err
		}
		s := string(data)
		c.TokenExchangeJSON = &s
	} else {
		c.TokenExchangeJSON = nil
	}

	// Persist the inline `oauth_config` block. Rehydrated by AfterFind so
	// the initiate-verification endpoint can feed it to InitiateOAuthFlow
	// the same way the UI Create handler does.
	if c.PendingOAuthConfig != nil {
		data, err := json.Marshal(c.PendingOAuthConfig)
		if err != nil {
			return err
		}
		s := string(data)
		c.PendingOAuthConfigJSON = &s
	} else {
		c.PendingOAuthConfigJSON = nil
	}

	// Encrypt sensitive fields after serialization.
	// Always set EncryptionStatus when encryption is enabled so the startup
	// batch pass does not re-process this row indefinitely.
	if encrypt.IsEnabled() {
		if c.ConnectionString != nil && !c.ConnectionString.IsFromSecret() && c.ConnectionString.GetValue() != "" {
			// Copy to avoid encrypting the shared ConnectionString through the pointer
			cs := *c.ConnectionString
			enc, err := encrypt.Encrypt(cs.Val)
			if err != nil {
				return fmt.Errorf("failed to encrypt mcp connection string: %w", err)
			}
			cs.Val = enc
			c.ConnectionString = &cs
		}
		if c.HeadersJSON != "" && c.HeadersJSON != "{}" {
			enc, err := encrypt.Encrypt(c.HeadersJSON)
			if err != nil {
				return fmt.Errorf("failed to encrypt mcp headers: %w", err)
			}
			c.HeadersJSON = enc
		}
		// The stash can carry an inline OAuth client_secret — encrypt it at
		// rest like the other credential-bearing columns.
		if c.PendingOAuthConfigJSON != nil && *c.PendingOAuthConfigJSON != "" {
			enc, err := encrypt.Encrypt(*c.PendingOAuthConfigJSON)
			if err != nil {
				return fmt.Errorf("failed to encrypt mcp pending oauth config: %w", err)
			}
			c.PendingOAuthConfigJSON = &enc
		}
		// The token-exchange block carries the exchange application's
		// client_secret — same treatment.
		if c.TokenExchangeJSON != nil && *c.TokenExchangeJSON != "" {
			enc, err := encrypt.Encrypt(*c.TokenExchangeJSON)
			if err != nil {
				return fmt.Errorf("failed to encrypt mcp token exchange config: %w", err)
			}
			c.TokenExchangeJSON = &enc
		}
		c.EncryptionStatus = EncryptionStatusEncrypted
	}

	return nil
}

// AfterFind is a GORM hook that decrypts the connection string and headers (if encrypted)
// and deserializes JSON columns back into runtime structs after reading from the database.
func (c *TableMCPClient) AfterFind(tx *gorm.DB) error {
	if c.EncryptionStatus == EncryptionStatusEncrypted {
		if c.HeadersJSON != "" && c.HeadersJSON != "{}" {
			decrypted, err := encrypt.Decrypt(c.HeadersJSON)
			if err != nil {
				return fmt.Errorf("failed to decrypt mcp headers: %w", err)
			}
			c.HeadersJSON = decrypted
		}
		if c.ConnectionString != nil && !c.ConnectionString.IsFromSecret() && c.ConnectionString.GetValue() != "" {
			decrypted, err := encrypt.Decrypt(c.ConnectionString.Val)
			if err != nil {
				return fmt.Errorf("failed to decrypt mcp connection string: %w", err)
			}
			c.ConnectionString.Val = decrypted
		}
		if c.TokenExchangeJSON != nil && *c.TokenExchangeJSON != "" {
			decrypted, err := encrypt.Decrypt(*c.TokenExchangeJSON)
			if err != nil {
				return fmt.Errorf("failed to decrypt mcp token exchange config: %w", err)
			}
			c.TokenExchangeJSON = &decrypted
		}
		if c.PendingOAuthConfigJSON != nil && *c.PendingOAuthConfigJSON != "" {
			decrypted, err := encrypt.Decrypt(*c.PendingOAuthConfigJSON)
			if err != nil {
				return fmt.Errorf("failed to decrypt mcp pending oauth config: %w", err)
			}
			c.PendingOAuthConfigJSON = &decrypted
		}
	}
	if c.StdioConfigJSON != nil {
		var config schemas.MCPStdioConfig
		if err := sonic.Unmarshal([]byte(*c.StdioConfigJSON), &config); err != nil {
			return err
		}
		c.StdioConfig = &config
	}
	if c.TLSConfigJSON != nil {
		var config schemas.MCPTLSConfig
		if err := sonic.Unmarshal([]byte(*c.TLSConfigJSON), &config); err != nil {
			return err
		}
		c.TLSConfig = &config
	}
	if c.ToolsToExecuteJSON != "" {
		if err := sonic.Unmarshal([]byte(c.ToolsToExecuteJSON), &c.ToolsToExecute); err != nil {
			return err
		}
	}
	if c.ToolsToAutoExecuteJSON != "" {
		if err := sonic.Unmarshal([]byte(c.ToolsToAutoExecuteJSON), &c.ToolsToAutoExecute); err != nil {
			return err
		}
	}
	if c.HeadersJSON != "" {
		if err := sonic.Unmarshal([]byte(c.HeadersJSON), &c.Headers); err != nil {
			return err
		}
	}
	if c.AllowedExtraHeadersJSON != "" {
		if err := sonic.Unmarshal([]byte(c.AllowedExtraHeadersJSON), &c.AllowedExtraHeaders); err != nil {
			return err
		}
	}
	if c.ToolPricingJSON != "" {
		if err := json.Unmarshal([]byte(c.ToolPricingJSON), &c.ToolPricing); err != nil {
			return err
		}
	}
	if c.DiscoveredToolsJSON != "" {
		if err := sonic.Unmarshal([]byte(c.DiscoveredToolsJSON), &c.DiscoveredTools); err != nil {
			return err
		}
	}
	if c.ToolNameMappingJSON != "" {
		if err := sonic.Unmarshal([]byte(c.ToolNameMappingJSON), &c.DiscoveredToolNameMapping); err != nil {
			return err
		}
	}
	if c.PerUserHeaderKeysJSON != "" {
		if err := sonic.Unmarshal([]byte(c.PerUserHeaderKeysJSON), &c.PerUserHeaderKeys); err != nil {
			return err
		}
	}
	if c.TokenExchangeJSON != nil && *c.TokenExchangeJSON != "" {
		var cfg schemas.MCPTokenExchangeConfig
		if err := sonic.Unmarshal([]byte(*c.TokenExchangeJSON), &cfg); err != nil {
			return err
		}
		c.TokenExchange = &cfg
	}
	if c.PendingOAuthConfigJSON != nil && *c.PendingOAuthConfigJSON != "" {
		var cfg schemas.OAuth2Config
		if err := sonic.Unmarshal([]byte(*c.PendingOAuthConfigJSON), &cfg); err != nil {
			return err
		}
		c.PendingOAuthConfig = &cfg
	}
	return nil
}

// VaultPathKey implements schemas.VaultPathKeyer so the global GORM vault
// callback can compute the vault base path for this model automatically.
func (c *TableMCPClient) VaultPathKey() string { return c.ClientID }

File: ui/lib/types/mcp.ts
import { Function as ToolFunction } from "./logs";
import { SecretVar } from "./schemas";

export type MCPConnectionType = "http" | "stdio" | "sse";

export type MCPConnectionState =
	| "healthy"
	| "unstable"
	| "error"
	| "pending_verification"
	| "disabled"
	| "needs_reauth"
	// Read-time aggregate value, never a single instance's own reported
	// state: multiple instances of a distributed deployment currently
	// report different states for this client. Never produced by a
	// single-instance deployment.
	| "degraded";

export type MCPAuthType = "none" | "headers" | "oauth" | "per_user_oauth" | "per_user_headers" | "token_exchange";

// Which step of Bifrost's own connection handling failed. Mirrors the Go
// MCPConnectionFailureStage constants.
export type MCPConnectionFailureStage = "connect" | "ping" | "list_tools" | "tool_discovery" | "transport_lost" | "credential";

// Why a server's state is not healthy: the step that failed, the error it
// failed with, the most recent failed attempt, and when the current run of
// failures began. Absent while healthy.
export interface MCPConnectionFailure {
	stage: MCPConnectionFailureStage;
	message: string;
	at: string;
	since: string;
}

// One instance's own view of a server in a distributed deployment.
export interface MCPInstanceState {
	state: MCPConnectionState;
	last_failure?: MCPConnectionFailure;
}

// Lifecycle states for a per-user MCP header credential row. Mirrors the
// status column on mcp_per_user_header_credentials.
//   - active:       caller-submitted, usable
//   - orphaned:     caller lost access via VK reassignment; auto-reactivates
//                   if access is regained
//   - needs_update: admin changed the PerUserHeaderKeys schema; caller must
//                   resubmit values
export type MCPHeadersUserCredentialStatus = "active" | "orphaned" | "needs_update";

export type { SecretVar };

export interface MCPStdioConfig {
	command: string;
	args: string[];
	envs: string[];
}

export interface MCPTLSConfig {
	insecure_skip_verify?: boolean;
	ca_cert_pem?: SecretVar;
}

// Delegated token-exchange scoping for auth_type === "token_exchange": each
// caller's identity-provider token is exchanged at runtime for a short-lived
// token scoped to this server's audience. Carries no credentials — the
// exchange endpoint and client come from the deployment's identity-provider
// integration.
export interface MCPTokenExchangeConfig {
	audience: string; // Resource identifier at the identity provider, e.g. "api://jira-mcp"
	// Identity-provider application authorized to perform exchanges for this
	// audience — typically a dedicated registration carrying the
	// token-exchange (or on-behalf-of) grant, separate from the SSO login
	// application. Required unless use_idp_credentials is true. Redacted on
	// GET.
	client_id?: SecretVar;
	client_secret?: SecretVar; // Omit for public clients. Ignored when use_idp_credentials is true. Redacted on GET.
	// When true, performs the exchange as the SSO login application itself
	// instead of client_id/client_secret above, which are then ignored. Some
	// providers require this — Microsoft Entra ID's on-behalf-of grant only
	// accepts an assertion audienced to the exchanging application, and the
	// SSO login flow always requests a token self-audienced to itself, so a
	// separate exchange application can never receive a usable one.
	use_idp_credentials?: boolean;
	// Optional scopes on the exchanged token. Include "offline_access" (where
	// the identity provider supports it) so the retained admin discovery
	// credential gets a refresh token and stays self-renewing.
	scopes?: string[];
	// Overrides where the exchange request is sent, for identity providers
	// that bind an audience to a specific authorization server distinct from
	// the one used for SSO login (e.g. Okta's per-resource Custom
	// Authorization Servers). Leave empty to use the deployment's SSO login
	// issuer, which is correct for providers with one tenant-wide token
	// endpoint (Entra, Auth0).
	authorization_server_url?: string;
}

export interface OAuthConfig {
	client_id: SecretVar;
	client_secret?: SecretVar; // Optional for public clients using PKCE
	authorize_url?: string; // Optional, will be discovered from server_url if not provided
	token_url?: string; // Optional, will be discovered from server_url if not provided
	registration_url?: string; // Optional, for dynamic client registration
	scopes?: string[]; // Optional, can be discovered
	server_url?: string; // MCP server URL for OAuth discovery (automatically set from connection_string)
	resource?: string; // Optional OAuth resource indicator; omitted when empty
}

/** OAuth fields allowed on MCP client update. Any field left unset keeps its stored value. */
export interface OAuthConfigUpdate {
	client_id?: SecretVar;
	client_secret?: SecretVar;
	authorize_url?: string;
	token_url?: string;
	registration_url?: string;
	scopes?: string[];
	resource?: string;
}

export interface MCPClientConfig {
	client_id: string; // Maps to ClientID in TableMCPClient
	name: string;
	endpoint_slug?: string; // URL-safe, immutable after creation; served at /mcp/<slug>
	is_code_mode_client?: boolean;
	connection_type: MCPConnectionType;
	connection_string?: SecretVar;
	stdio_config?: MCPStdioConfig;
	tls_config?: MCPTLSConfig;
	auth_type?: MCPAuthType;
	oauth_config_id?: string;
	oauth_client_id?: SecretVar; // Redacted existing client ID (populated on GET for oauth clients)
	oauth_client_secret?: SecretVar; // Redacted existing client secret (populated on GET for oauth clients)
	// Remaining oauth_config fields, populated on GET for oauth clients so the
	// full config can be reviewed and edited, not just the credentials.
	oauth_authorize_url?: string;
	oauth_token_url?: string;
	oauth_registration_url?: string;
	oauth_scopes?: string[];
	oauth_resource?: string;
	tools_to_execute?: string[];
	tools_to_auto_execute?: string[];
	headers?: Record<string, SecretVar>;
	// per_user_header_keys: admin-declared header *names* that each caller
	// must supply when auth_type === "per_user_headers". Values live per-user
	// in the credential store, not on the client config. Required (non-empty)
	// for per_user_headers auth; ignored for all other auth types.
	per_user_header_keys?: string[];
	// token_exchange-only: audience/scopes the exchanged tokens are scoped to.
	// Required (with a non-empty audience) for token_exchange auth.
	token_exchange?: MCPTokenExchangeConfig;
	is_ping_available?: boolean;
	// Only meaningful when connection_type === "http": whether this client
	// maintains one persistent connection reused across every caller (true)
	// or connects fresh per call (nil/false, the default for newly created
	// clients — pre-existing clients were backfilled to true). SSE and STDIO
	// always behave as sticky regardless of this field.
	needs_session_stickiness?: boolean;
	tool_pricing?: Record<string, number>;
	// Per-client override (0 = use global). API returns NANOSECONDS
	// (Go time.Duration), while updates send minutes — convert with
	// toolSyncIntervalToMinutes before showing or resending this value.
	tool_sync_interval?: number;
	tool_execution_timeout?: string | number; // Per-client tool execution timeout; API returns string e.g. "30s", UI sends integer seconds (0 = use global)
	allowed_extra_headers?: string[]; // Allowlist of x-bf-eh-* headers forwarded to this MCP server. ["*"] = allow all.
	allow_by_default?: boolean; // When true, available to every caller not assigned this server explicitly, with all tools allowed; an explicit assignment overrides this
	disabled?: boolean; // When true, connection/workers are shut down; tools are unavailable until re-enabled
}

export interface MCPVKConfigResponse {
	virtual_key_id: string;
	virtual_key_name: string;
	tools_to_execute: string[];
}

// MCPClientCredential is the read-only view of the credential a server holds
// on its own behalf, as opposed to the per-caller credentials listed under
// MCP sessions: the single shared token for an oauth server, the retained
// admin token for per_user_oauth and token_exchange servers (used only to
// refresh the tool list), or the retained admin header values for a
// per_user_headers server. Never carries secret material: the refresh token
// is reported as present or absent, and header values only by name.
export interface MCPClientCredential {
	kind: "oauth" | "headers";
	// oauth: active | orphaned | needs_reauth. headers: active | orphaned | needs_update.
	status: "active" | "orphaned" | "needs_reauth" | "needs_update";
	// oauth only: why the status left active, as recorded when it flipped:
	// the provider's rejection of the last refresh (HTTP status plus the
	// OAuth error and description), a credential rotation, or a failed admin
	// exchange. Absent while active.
	status_reason?: string;
	// oauth only: access token expiry; absent when the provider did not report one.
	expires_at?: string | null;
	// oauth only: set once the access token has been refreshed or the
	// credential re-authorized at least once.
	last_refreshed_at?: string | null;
	// oauth only: whether the provider issued a refresh token. Always false for headers.
	has_refresh_token: boolean;
	// oauth only: scopes the provider reported at consent. Empty when the
	// provider omitted them from its token response.
	scopes?: string[];
	// headers only: names of the headers the stored admin values cover, sorted.
	header_keys?: string[];
	created_at: string;
	updated_at: string;
}

export interface MCPClient {
	config: MCPClientConfig;
	tools: ToolFunction[];
	state: MCPConnectionState;
	vk_configs: MCPVKConfigResponse[];
	// The failure behind a `state` other than healthy, as recorded by the
	// instance that served this request. Absent while healthy.
	last_failure?: MCPConnectionFailure;
	// Per-instance breakdown behind `state` in a distributed deployment
	// (instance ID -> that instance's own state and failure). Present when
	// instances disagree (`state` is then "degraded") and when they all
	// report unstable, so each instance's own reason is visible. Never
	// present in a single-instance deployment.
	node_states?: Record<string, MCPInstanceState>;
	// The credential this server holds on its own behalf. Absent for auth
	// types without one (none, headers) and for servers that have not
	// completed their one-time authorization yet.
	credential?: MCPClientCredential;
}

export interface CreateMCPClientRequest {
	name: string;
	endpoint_slug?: string; // Optional on create (derived from name when blank); immutable after
	is_code_mode_client?: boolean;
	connection_type: MCPConnectionType;
	connection_string?: SecretVar;
	stdio_config?: MCPStdioConfig;
	tls_config?: MCPTLSConfig;
	auth_type?: MCPAuthType;
	oauth_config?: OAuthConfig;
	tools_to_execute?: string[];
	tools_to_auto_execute?: string[];
	headers?: Record<string, SecretVar>;
	// per_user_headers-only: admin-declared header schema (names only).
	per_user_header_keys?: string[];
	// token_exchange-only: audience/scopes the exchanged tokens are scoped to.
	// Verification runs automatically as the signed-in admin (their own
	// identity-provider token is the exchange subject); sessions without one
	// (e.g. API-key auth) create the client in pending_verification.
	token_exchange?: MCPTokenExchangeConfig;
	// per_user_headers-only: a sample set of header values supplied by the
	// admin so the server can verify upstream + discover tools in the same
	// create call. Discarded after verification (never persisted). Mirrors
	// the per-user OAuth flow where the admin's temp access token plays
	// the analogous role. Ignored for all other auth types.
	user_headers?: Record<string, string>;
	is_ping_available?: boolean;
	// Only meaningful when connection_type === "http". See MCPClientConfig's
	// field doc for the full contract.
	needs_session_stickiness?: boolean;
}

export interface OAuthFlowResponse {
	status: "pending_oauth";
	message: string;
	oauth_config_id: string;
	authorize_url: string;
	expires_at: string;
	mcp_client_id: string;
}

export interface OAuthStatusResponse {
	id: string;
	status: "pending" | "authorized" | "failed" | "expired" | "revoked";
	created_at: string;
	token_id?: string;
	token_expires_at?: string;
	token_scopes?: string;
}

export interface MCPVKConfig {
	virtual_key_id: string;
	tools_to_execute: string[];
}

export interface UpdateMCPClientRequest {
	name?: string;
	is_code_mode_client?: boolean;
	headers?: Record<string, SecretVar>;
	// Set to a new list (including empty) to replace per-user-headers schema.
	// Omitted = preserve existing. When this list changes against the stored
	// value, the backend flips all existing user credential rows to
	// 'needs_update' so callers re-submit on next tool use.
	per_user_header_keys?: string[];
	tools_to_execute?: string[];
	tools_to_auto_execute?: string[];
	is_ping_available?: boolean;
	// Only meaningful when connection_type === "http". Toggling this on an
	// existing client takes effect immediately: switching to true dials the
	// persistent connection now, switching to false closes it. See
	// MCPClientConfig's field doc for the full contract.
	needs_session_stickiness?: boolean;
	tool_pricing?: Record<string, number>;
	tool_sync_interval?: number; // Per-client override in minutes (0 = use global)
	tool_execution_timeout?: number; // Per-client tool execution timeout in seconds (0 = use global)
	allowed_extra_headers?: string[]; // Allowlist of x-bf-eh-* headers forwarded to this MCP server. ["*"] = allow all.
	allow_by_default?: boolean; // When true, available to every caller not assigned this server explicitly, with all tools allowed; an explicit assignment overrides this
	disabled?: boolean; // Set to true to shut down connection/workers; false to reconnect
	tls_config?: MCPTLSConfig; // TLS configuration for HTTP/SSE connections
	oauth_config?: OAuthConfigUpdate; // Only supported for existing oauth/per_user_oauth clients (credential rotation)
	token_exchange?: MCPTokenExchangeConfig; // Only supported for existing token_exchange clients; omitted = preserve
	vk_configs?: MCPVKConfig[]; // When provided, replaces all VK assignments for this MCP client
}

// Pagination + filter params for MCP clients list
export interface GetMCPClientsParams {
	limit?: number;
	offset?: number;
	search?: string;
	server?: string;
	// Comma-separated exact-match filters (OR semantics), mirroring the library page.
	connection_type?: string; // http,sse,stdio
	auth_type?: string; // none,headers,oauth,per_user_oauth,per_user_headers,token_exchange
	state?: string; // connected,disconnected — resolved against live engine state
	virtual_keys?: string; // comma-separated VK IDs the client is assigned to
	// Boolean facets — omit for "no filter".
	code_mode?: boolean; // filters is_code_mode_client
	disabled?: boolean; // filters disabled status
	allowed_by_default?: boolean; // when true, include clients allowed by default
}

// Paginated response for MCP clients list
export interface GetMCPClientsResponse {
	clients: MCPClient[];
	count: number;
	total_count: number;
	limit: number;
	offset: number;
}

// Types for MCP Tool Selector component
export interface SelectedTool {
	mcpClientId: string;
	toolName: string;
}

// MCP Tool Spec for tool groups (matches backend schema)
export interface MCPToolSpec {
	mcp_client_id: string;
	tool_names: string[];
}

// ---------------------------------------------------------------------------
// MCP Library (synced catalog)
// ---------------------------------------------------------------------------

/** A single entry from the synced MCP server catalog (`mcp_library` table). */
export interface MCPLibraryEntry {
	id: number;
	slug: string;
	name: string;
	description?: string;
	category?: string;
	connection_type: MCPConnectionType;
	connection_url?: string;
	stdio_config?: MCPStdioConfig;
	auth_type?: MCPAuthType;
	required_header_keys?: string[];
	icon_url?: string;
	docs_url?: string;
	publisher?: string;
	tags?: string[];
	metadata?: Record<string, unknown>;
	/** "remote" for synced rows, "custom" for org-published entries. */
	source?: "remote" | "custom";
	created_at: string;
	updated_at: string;
}

/** Body for POST /api/mcp/library — publish a custom (org-internal) library entry. */
export interface CreateMCPLibraryEntryRequest {
	name: string;
	description?: string;
	category?: string;
	connection_type: MCPConnectionType;
	connection_url?: string;
	stdio_config?: MCPStdioConfig;
	auth_type?: MCPAuthType;
	required_header_keys?: string[];
	icon_url?: string;
	docs_url?: string;
	publisher?: string;
	tags?: string[];
}

export interface GetMCPLibraryParams {
	search?: string;
	category?: string;
	connection_type?: string;
	auth_type?: string;
	tags?: string;
	sort_by?: string;
	order?: string;
	limit?: number;
	offset?: number;
}

export interface GetMCPLibraryResponse {
	servers: MCPLibraryEntry[];
	count: number;
	total_count: number;
	limit: number;
	offset: number;
}

export interface MCPLibraryFilterData {
	categories: string[];
	connection_types: string[];
	auth_types: string[];
	tags: string[];
}